### ISSUE TITLE
Add support for Project as a configuration node

### ISSUE THREAD

**Issue by cdonati** [GoogleCloudPlatform/kafka-pubsub-emulator#6]
There seems to be no way to prevent requests to "project-a/topic-1" from being treated the same as requests to "project-b/topic-1".

My understanding is that the character set allowed by Kafka for topic names prevents automatic namespacing just based on the Kafka topic name. However, it would be useful to be able to explicitly define project IDs for each topic and subscription in the emulator configuration.

**Comment by prodonjs** [comment]
Thanks Chris. Assuming that you are OK with having one Kafka cluster reflect multiple projects, we could add support for namespaced Cloud Pub/Sub topic names that map to valid Kafka topic names. On the Kafka side, this would make it more challenging to map topic names back to their CPS topics, but it would support the logical separation for CPS clients. Unless we were to change our configuration and connection model, it would not be possible to have one emulator support topics from different Kafka clusters.

I'm open to consider whichever option you think is most tenable.

**Comment by cdonati** [comment]
Personally, I would be fine with having one Kafka cluster handling multiple projects. I think that would be the best fit as far as AppScale is concerned. A single deployment would only use one Kafka cluster, but it would host multiple applications.

**Comment by cdonati** [comment]
Thank you for adding the project bind configuration. From the configuration example, it appears that the Cloud Pub/Sub topic name still needs to be the same as the Kafka topic name. For example, I would not be able to create a Kafka topic named "project-a_topic-a" and bind a Pub/Sub "project-a/topic-a" to that topic. Am I understanding that correctly?

**Comment by prodonjs** [comment]
You are correct. We'll take another pass.

**Comment by prodonjs** [comment]
Sorry Chris, I don't think I was clear when I talked with Iago about this. Here is what we are thinking now.

If we added another optional property to the pubsub project section in our configuration like this:

```
pubsub:
  project-a:
    - topic: topic-1
      kafka-topic: project-a-topic-1
      subscription: subscription-to-topic-1
  project-b:
    - topic: topic-1
      kafka-topic: project-b-topic-1
      subscription: subscription-to-topic-1
```

We would then be able to publish to `projects/project-a/topics/topics-1` and `projects/project-b/topics/topics-1` and the emulator would lookup the Kafka topic to determine the appropriate place to publish to. If the `kafka-topic` section was not populated, we'd use `topic`. The same logic would apply for Subscriptions.

Does that seem reasonable?

**Comment by cdonati** [comment]
No worries. I was not very specific about the functionality I was looking for. Thank you for your help.

That syntax looks like it would work perfectly as far as topics go.

I'm still a bit uncertain about how it would work for subscriptions. In the example you gave, what would go in the kafka.consumer.subscriptions section? Would there be multiple subscriptions with the same name? For example:
```
subscriptions:
 - name: subscription-to-topic-1
   topic: project-a-topic-1
 - name: subscription-to-topic-1
   topic: project-b-topic-1
```

Or would the user omit the kafka.consumer.subscriptions section if they were explicitly defining the subscriptions in the pubsub section?

Named subscriptions seem to be specific to the Pub/Sub API rather than a Kafka detail, so it would make sense to me to drop the section from the kafka stanza and use the pubsub one to define subscriptions. However, that would require the user to assign a project to each subscription. While that would be fine for AppScale's usage (whenever we create a topic or sub, we'd know which project it would be for), you may still be interested in supporting a more simplified configuration without project IDs.

One last thing I'm unclear about is how to define Pub/Sub subscriptions when you don't have exactly one per topic. I assume that when a topic does not have a subscription, the user would omit the subscription key from the block. What would the user do if they have multiple subscriptions for a given Pub/Sub topic?

**Comment by prodonjs** [comment]
Yes so I'm thinking we would drop the `producer.topics` and `consumer.subscriptions` sections entirely. All of the Pub/Sub specific configuration would reside in the `pubsub` section.

Both the REST and gRPC APIs require Topic and Subscription names to be fully qualified, so I don't see an issue with mandating that a configuration need to introduce a project ID. It does make the configuration a bit more terse than before, but then we're being explicit about how we handle the topic and subscription names in our code instead of just dropping the prefixes as we are now.

For a Topic with multiple Subscriptions, we can just change our YAML syntax to support an optional list of Subscription names.

```
pubsub:
  project-a:
    - topic: topic-1
      kafka-topic: project-a-topic-1
      subscriptions: 
        - subscription-1-to-topic-1
        - subscription-2-to-topic-1
    - topic: topic-2
      kafka-topic: project-a-topic-2
  project-b:
    - topic: topic-1
      kafka-topic: project-b-topic-1
      subscriptions: 
        - subscription-to-topic-1
```

**Comment by cdonati** [comment]
Ah, I see. Yes, that looks great!

**Comment by prodonjs** [comment]
#15 Provides support for multiple projects via a new JSON-based configuration format. See https://github.com/GoogleCloudPlatform/kafka-pubsub-emulator/blob/v0.1.0/demo/benchmark/config/config.json for an example.

This code will be pushed to master once all supporting documentation is updated.

**Comment by prodonjs** [comment]
This issue is now closed by PR #17
